### PR TITLE
Buffered Async Block Stream

### PR DESCRIPTION
**Pull request by ewoolsey**
This PR improves the Block scanner using async streams, allowing for buffered requests greatly increasing throughput and sync speed. Additionally adds some retry mechanisms for failed network requests.

### CODE REVIEW

File: src/tree/block_scanner.rs [hunk 1]
```diff
@@ -27,47 +34,77 @@ where
...
             };
 
-        let filter = self
-            .filter
-            .clone()
-            .from_block(BlockNumber::Number((*next_block).into()))
-            .to_block(BlockNumber::Number(to_block.into()));
+            let filter = self
+                .filter
+                .clone()
```
Comment (0xKitsune) [inline src/tree/block_scanner.rs:81]: Consider simplifying to one match arm.

```suggestion
                Ok(latest) => {  
                        if latest.as_u64() < next_block {  
                            tokio::time::sleep(Duration::from_secs(  
                                BLOCK_SCANNER_SLEEP_TIME,  
                            ))  
                            .await;  
                            continue;  
                        } else{  
                            break (next_block + self.window_size)  
                                .min(latest.as_u64());  
                        }  
                        
                    }  
```

File: src/tree/block_scanner.rs [hunk 2]
```diff
@@ -27,47 +34,77 @@ where
...
 
-        let middleware = self.middleware.clone();
+            let middleware = self.middleware.clone();
+            let chain_id = self.chain_id;
 
-        let logs = middleware.get_logs(&filter).await?;
+            // This future is yielded from the stream
+            // and is awaited on by the caller
+            let fut = retry(
+                Duration::from_millis(100),
```
Comment (0xKitsune) [inline src/tree/block_scanner.rs:93]: Can we add error logging when this happens?

File: src/tree/block_scanner.rs [hunk 3]
```diff
@@ -27,47 +34,77 @@ where
...
-            } else {
-                return Ok((0, vec![]));
+    pub fn block_stream(
+        &self,
+    ) -> impl Stream<Item: Future<Output = WorldTreeResult<Vec<Log>>> + Send> + '_
+    {
+        stream::unfold(self.start_block, move |mut next_block| async move {
+            // This is executed before the item is yielded
+            let to_block = loop {
+                match self.middleware.get_block_number().await {
```
Comment (0xKitsune) [inline src/tree/block_scanner.rs:57]: Previously we have seen connection closed errors occasionally. IIRC ethers does not have reconnect logic under the hood. In the case that the connection closes, would we perpetually loop here?

File: src/tree/tree_manager.rs [hunk 4]
```diff
@@ -104,30 +97,21 @@ impl TreeVersion for CanonicalTree {
...
-                        .await;
-
-                        return ok(());
-                    }
-
+            block_scanner
+                .block_stream()
+                // Retrieve logs concurrently
+                // Setting this too high can cause a 502
+                .buffered(4)
```
Comment (0xKitsune) [inline src/tree/tree_manager.rs:111]: Will `4` be enough to sync on mainnet?

File: src/tree/block_scanner.rs (hunk 1 above)
Comment (ewoolsey) [reply src/tree/block_scanner.rs:81]: Not a fan of `if` inside match arms eh?

File: src/tree/block_scanner.rs (hunk 3 above)
Comment (ewoolsey) [reply src/tree/block_scanner.rs:57]: Yes it's possible we perpetually loop here. How would you fix that though, a limited number of retries?

File: src/tree/block_scanner.rs (hunk 1 above)
Comment (0xKitsune) [reply src/tree/block_scanner.rs:81]: Either works, a single match arm is a bit more succinct here. 

If you feel strongly feel free to leave it!

File: src/tree/tree_manager.rs (hunk 4 above)
Comment (ewoolsey) [reply src/tree/tree_manager.rs:111]: I'm not sure what you mean. The 4 here means we can request blocks 4 times concurrently. Previously this was effectively less than 1.

File: src/tree/block_scanner.rs (hunk 3 above)
Comment (0xKitsune) [reply src/tree/block_scanner.rs:57]: Yes we could potentially exit after a specified number of retries. We could also pass the rpc url  to the block scanner and try to reconnect after repeated failures.

File: src/tree/tree_manager.rs (hunk 4 above)
Comment (0xKitsune) [reply src/tree/tree_manager.rs]: I misread, disregard!

**Review by 0xKitsune — APPROVED** [review]

**Pull request merged**